blog/2023-11-09-useState-and-useEffect-hooks/index.mdx

```diff
@@ -0,0 +1,265 @@
+
+Once the state value is updated, the component is rerendered, executing the body of the function component again, and recreating the rendered HTML.
+But the state is unaffected, hence the value of `count` is maintained, and the function reference `setCount` remains unchanged.
+
+### Initialising the state
+
+The `useState` method is generic, allowing for any valid type to be used.
+
+There are two methods for initialising the state. Regardless of the approach, the state that is returned will be equal to the initial value.
+
+The first approach is to provide the initial value directly to the `useState` hook call.
+This is appropriate when the initial value is known, for example knowing the counter will start from 0.
+
+```tsx showLineNumbers
+import {useState} from "react";
+
+function ButtonCounter() {
+  const [count, setCount] = useState<number>(0);
+  // ...
+}
+```
+
+The alternative is to provide a function which takes no arguments and returns the initial value.
+This is appropriate when the initial value must perform an intensive computation as this function is only called on the first render.
+
+Taking this approach avoids the recomputation of the initial value after the first render as it will no longer be required.
+
+```tsx showLineNumbers
+import {useState} from "react";
+
+function getInitialCount() {
+  let count = 0;
+  // Intensive computation
+  return count
+}
+
+function ButtonCounter() {
+  /* 💣 Do not do this */
+  // const initialValue = getInitialCount()
+  // const [count, setCount] = useState<number>(initialValue);
+
+  const [count, setCount] = useState<number>(() => getInitialCount());
+  // ...
+}
+```
+
+### setState
+
+The useState hook returns an array that is two elements long, the first being the initial value, and the second being a function to set the state, e.g.
+
+```ts
+const [state: T, setState: (prev: T) => T] = useState<T>(initialValue: T | getInitialValue: () => T);
+```
+
+The `setState` function can be provided the value for the next state, or a function which takes the previous state and returns the next state.
+If setting the state uses the previous state, then you should be providing a function to avoid any synchronisation errors.
+This means the component can also be written.
+
+```tsx showLineNumbers
+import { useState } from "react";
+
+function ButtonCounter() {
+  const [count, setCount] = useState<number>(0);
+
+  const incrementCounter = () => {
+    setCount(c => c + 1);
+  }
+
+  return (
+      <>
+        <button onClick={() => incrementCounter()}>Click me to increase counter!</button>
+        <p>Times pressed: {count}</p>
+      </>
+  );
+}
+```
+
+## useEffect
+
+The `useEffect` hook allows side effects to be performed within functional components.
+This removes the requirement for developers to write asynchronous code.
+In particular, there are three main variations that are commonly seen.
+
+### Return value
```

Comment:
seq: be a newcomer

1. start without knowledge - no deps array
2. deps array
3. cleanup